database/connect.ts

```diff
@@ -32,7 +32,12 @@ function connectOneTimeToDatabase() {
     });
   }
 
-  return globalThis.postgresSqlClient;
+  return <T extends (object | undefined)[]>(
+    ...sqlQuery: [TemplateStringsArray, ...any[]]
+  ) => {
+    headers();
```

Comment:
Calling headers() in every sql Query will prevent Next.js wrongly flag pages as Static Rendering. Preventing the extra need of configuration. A newhttps://github.com/vercel/next.js/discussions/50695#discussion-5258288to avoid the need of this function call.

When calling a Dynamic Function as headers() in sql client there is no need to write this line since [Next.js will assume this is a Dynamic Rendered page](https://nextjs.org/docs/app/building-your-application/rendering/static-and-dynamic-rendering#dynamic-rendering)

Reply:
Alternatives considered:

1. call `headers()` at the top level of the `database/connect.ts` function (downside: I didn't get this working yet, because of the `Method expects to have requestAsyncStorage, none available` below)
2. call `headers()` in every database query function and not modify the client call (downside: verbosity, lots of changes, continued maintenance effort)
3. call every database function from API Route Handlers that allow the revalidate option to be used to prevent Static Rendering (downside: architectural changes, lots of changes, overhead)

```
info Collecting page data ..Error: Invariant: Method expects to have requestAsyncStorage, none available
    at headers (/Users/josehower/projects/next-js-example-spring-2023-vienna-austria/.next/server/chunks/647.js:89:15)
    at 7147 (/Users/josehower/projects/next-js-example-spring-2023-vienna-austria/.next/server/chunks/147.js:34:20)
    at __webpack_require__ (/Users/josehower/projects/next-js-example-spring-2023-vienna-austria/.next/server/webpack-runtime.js:25:43)
    at 9040 (/Users/josehower/projects/next-js-example-spring-2023-vienna-austria/.next/server/app/animal-management-naive-dont-copy/delete/[animalId]/page.js:339:75)
    at Function.__webpack_require__ (/Users/josehower/projects/next-js-example-spring-2023-vienna-austria/.next/server/webpack-runtime.js:25:43)
```

Reply:
by @karlhorky 

options 2 and 3 have significant downsides as you mention, so probably good to avoid those

let's dig in more to different options that don't involve wrapping the `postgres()` function, including:

A) Option 1. above - try it out again and document your experience / code and any errors and their error messages in the discussion: https://github.com/vercel/next.js/discussions/50695

B) any other place that we can put `headers()`

the `requestAsyncStorage` problem may be something that we can work around - there is an issue here that has a similar-sounding error message:

- https://github.com/vercel/next.js/issues/50099

Reply:
Update: @Eprince-hub investigated Option A1) above (top-level `headers()` in the `connect.ts` file) but it didn't work

Some other options include:

A2) Create a new wrapper to replace `cache()` in the `util/` folder - could be something like `query()` or something. (downsides: this adds more files changed)

A3) Patch `next` (downside: again we need to deliver and maintain the patch)

but these are significant downsides for a hack / workaround. So let's go with @Josehower's solution in this PR

Reply:
https://github.com/upleveled/next-js-example-spring-2023-vienna-austria/pull/3/commits/9ac02405dcc22b7ea7ead18604e51fea63cc8493

Reply:
https://github.com/upleveled/next-js-example-spring-2023-vienna-austria/pull/3/commits/0413ec310986a7b61736de190c7bef8303c6be19
https://github.com/upleveled/next-js-example-spring-2023-vienna-austria/pull/3/commits/9ac02405dcc22b7ea7ead18604e51fea63cc8493